core/Tracker/Request.php

```diff
@@ -86,12 +86,25 @@ public function __construct($params, $tokenAuth = false)
             }
         }
 
-        // check for 4byte utf8 characters in url and replace them with �
+        // check for 4byte utf8 characters in all tracking params and replace them with �
         // @TODO Remove as soon as our database tables use utf8mb4 instead of utf8
-        if (array_key_exists('url', $this->params) && preg_match('/[\x{10000}-\x{10FFFF}]/u', $this->params['url'])) {
-            Common::printDebug("Unsupport character detected. Replacing with \xEF\xBF\xBD");
-            $this->params['url'] = preg_replace('/[\x{10000}-\x{10FFFF}]/u', "\xEF\xBF\xBD", $this->params['url']);
+        $this->params = $this->replaceUnsupportedUtf8Chars($this->params);
```

Comment:
Just to understand... I'm testing it right now and wondering why we need the replace here and below in `getParam()`? I understand in `getParam` might be because it returns an array from json? Is it still needed here as well? Is this in case someone does `->getParams()`?

Reply:
Yes. That was meant for `getParams()`. Not the perfect solution. But as it will be removed in Matomo 4 again, guess it's "good enough".

Reply:
OK was just wondering re performance etc but shouldn't be big.